crates/sui-core/src/authority_active/checkpoint_executor/mod.rs

```diff
@@ -0,0 +1,236 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{sync::Arc, time::Duration};
+
+use futures::stream::FuturesOrdered;
+use sui_metrics::spawn_monitored_task;
+use sui_types::{
+    base_types::{AuthorityName, ExecutionDigests, TransactionDigest},
+    committee::StakeUnit,
+    error::{SuiError, SuiResult},
+    messages::{TransactionEffects, VerifiedCertificate},
+    messages_checkpoint::{CheckpointSequenceNumber, VerifiedCheckpoint},
+};
+use tokio::{
+    sync::broadcast::{self, error::RecvError},
+    task::JoinHandle,
+};
+use tokio_stream::StreamExt;
+use tracing::{error, info, warn};
+use typed_store::rocks::TypedStoreError;
+
+use crate::{
+    authority::{AuthorityState, EffectsNotifyRead},
+    checkpoints::{CheckpointMetrics, CheckpointStore},
+};
+
+const TASKS_PER_CORE: usize = 1;
+
+pub struct CheckpointExecutor {
+    mailbox: broadcast::Receiver<VerifiedCheckpoint>,
+    checkpoint_store: Arc<CheckpointStore>,
+    authority_state: Arc<AuthorityState>,
+    highest_executed_checkpoint: Option<CheckpointSequenceNumber>,
+    metrics: Arc<CheckpointMetrics>,
+}
+
+impl CheckpointExecutor {
+    pub fn new(
+        state_sync_handle: &sui_network::state_sync::Handle,
+        checkpoint_store: Arc<CheckpointStore>,
+        authority_state: Arc<AuthorityState>,
+        metrics: Arc<CheckpointMetrics>,
+    ) -> Result<Self, TypedStoreError> {
+        let mailbox = state_sync_handle.subscribe_to_synced_checkpoints();
```

Comment:
We need to understand why is subscribing here safe in terms of startup race conditions. If state sync is already started at this point, could it be that some checkpoints were already processed and sent to channel before we subscribe here? (in which case they will be lost)

Reply:
If state sync has already started, and there are checkpoints that are synced before we subscribe, then it is as you say - the new checkpoint message will be lost. But even if the message is lost, the checkpoint will eventually be executed. On each message receipt, `CheckpointExecutor` begins syncing from the last executed up to the last synced, so it will be handled then. 

That said, you bring up a good point - `CheckpointExecutor` should probably check how far behind it is from the synced watermark at startup rather than wait for the next checkpoint message. Will add something for that